### PR TITLE
fmt: `deno fmt -` formats stdin and print to stdout

### LINKED ISSUE #3877: `deno fmt` should support read js/ts code from `stdin`
support read code from `stdin` and output to `stdout`

```shell
cat example.ts | deno fmt
```

Before replace with `dprinf`, `std/prettier` supports reading code from stdin

### Why?

I am developing [axetroy.vscode-deno](https://github.com/axetroy/vscode-deno)

To implement a formatting tool, it needs to be unrelated to the file system.

vscode file content is stored in memory before saving the file (same with others editor)

If you read the code from the file system and format it, it will cause a conflict

For it to work properly, it must be save in the file system before formatting.

**All we need is a file system-independent formatting tool.**

This problem is not only with vscode, it is the same in other editors

### PR DESCRIPTION
**Pull request by kevinkassimo**
Closes #3877

~~Add `--stdin` flag (also compatible with `--check`) to format stdin input~~

`deno fmt -` now reads from stdin and prints output to stdout.
```
$ echo "const    a=1" | ./target/debug/deno fmt -
const a = 1;
```

### CODE REVIEW

**Comment by nayeemrmn** [comment]
I think it's common to use `-` in place of the file name to represent reading from `stdin` / writing to `stdout`.

**Comment by kevinkassimo** [comment]
@nayeemrmn That’s also an option. Love to get more feedback also from others

**Comment by dplewis** [comment]
Would it be possible to achieve this without the flag. If no files input then read from pipe?

**Comment by axetroy** [comment]
@dplewis 

agree. See what I wrote before

https://github.com/denoland/deno/blob/8bc639a23e141d350e35ea112a4a9be2ea536fe6/std/prettier/main.ts#L548-L552

**Comment by nayeemrmn** [comment]
> How about without the flag? If no files input then read from pipe?

Hmm, we may want to have a default glob for this case.

**Comment by kevinkassimo** [comment]
@dplewis Currently if we don't have any arguments, we implicitly push `**/*`
https://github.com/denoland/deno/blob/f650c3edb3081ac0433b338c0c99418171656dd1/cli/fmt.rs#L162

**Comment by dplewis** [comment]
Sorry I'm a beginner at rust 😅If there are no arguments couldn't you check stdin().lines() first before pushing **/*?

File: cli/flags.rs [hunk 1]
```diff
@@ -555,6 +558,12 @@ fn fmt_subcommand<'a, 'b>() -> App<'a, 'b> {
         .help("Check if the source files are formatted.")
         .takes_value(false),
     )
+    .arg(
+      Arg::with_name("stdin")
+        .long("stdin")
+        .help("Read the code from stdin.")
```
Comment (bartlomieju) [inline cli/flags.rs]: It looks like besides reading from stdin it outputs code to stdout. Can you describe this behavior in help text? (Maybe add an example to subcommand as well)

**Comment by bartlomieju** [comment]
> @dplewis Currently if we don't have any arguments, we implicitly push `**/*`
> https://github.com/denoland/deno/blob/f650c3edb3081ac0433b338c0c99418171656dd1/cli/fmt.rs#L162

@ry and I had a talk about it - we're not happy with current glob solution and are thinking about removing it entirely for now. Glob crates available in Rust have completely different functionality (and it seems syntax too) than `std/globrex` that we used before.

File: cli/fmt.rs [hunk 2]
```diff
@@ -170,3 +174,31 @@ pub fn format_files(maybe_files: Option<Vec<String>>, check: bool) {
...
+pub fn format_stdin(check: bool) {
+  let mut source = String::new();
+  if stdin().read_to_string(&mut source).is_err() {
+    eprintln!("Failed to read from stdin");
+  }
+  let config = get_config();
+
+  match dprint::format_text("_stdin.ts", &source, &config) {
+    Ok(None) => {
+      // Should not happen, but whatever.
```
Comment (bartlomieju) [inline cli/fmt.rs]: `unreachable!()`?

File: cli/flags.rs (hunk 1 above)
Comment (nayeemrmn) [reply cli/flags.rs]: This is another reason `-` would be better. The meaning of this functionality is specifically "`stdin`/`stdout`". This is conventionally represented with `-`, whereas `--stdin` is misleading. I feel more strongly about it now.

Comment (kevinkassimo) [reply cli/flags.rs]: @nayeemrmn `-` can be a valid filename. (nah I'm just joking + bikeshedding here)

Yeah `--stdin` is a bit long to type and its implicit effect on stdout is definitely slightly confusing... If @bartlomieju also thinks `-` is better then it is a very easy change to make.

Comment (bartlomieju) [reply cli/flags.rs]: Yeah, SGTM, if that's an easy change I guess it makes more sense - we already have similar argument in `catj`.

**Pull request merged**